src/frontend/src/optimizer/plan_node/generic/cte_ref.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::hash::Hash;
+
+use itertools::Itertools;
+use pretty_xmlish::{Pretty, StrAssocArr};
+use risingwave_common::catalog::Schema;
+
+use super::{impl_distill_unit_from_fields, GenericPlanNode, GenericPlanRef};
+use crate::binder::ShareId;
+use crate::optimizer::property::FunctionalDependencySet;
+use crate::{optimizer, OptimizerContextRef};
+
+#[derive(Clone, Debug)]
+pub struct CteRef<PlanRef> {
+    share_id: ShareId,
+    base: PlanRef,
```

Comment:
IIUC, this `base` shouldn't be used after planning, right?

Reply:
the `base` here refers to the left-hand side base case for a rcte, and the use here is just to fetch the `schema` and the `OptimizerContextRef` - since `plan_recursive_union` will finish after `plan_cte_ref`.

I'm not sure what you exactly mean by "shouldn't be used after planning", but at least the methods for impl `GenericPlanRef` need this `base`.

Reply:
I mean once we have `rcte_cache_plan`, it should be always true to use `rcte_cache_plan`, right? Maybe we should add an assertion to ensure schema, stream_key and functional_dependency are the same between `rcte_cache_plan` and the `base`.

Reply:
> I mean once we have rcte_cache_plan, it should be always true to use rcte_cache_plan

yep, problem (and reason for `base` here) is we need to handle the gap between `plan_cte_ref` and `plan_recursive_union`.